Compilation failed
```
[INFO] compiling 2 Scala sources and 101 Java sources to /home/runner/work/lance-spark/lance-spark/lance-spark-4.1_2.13/target/test-classes ...
Error:  /home/runner/work/lance-spark/lance-spark/lance-spark-base_2.12/src/test/java/org/lance/spark/read/BaseSpjBroadcastTest.java:81:17:  error: unreported exception IOException; must be caught or declared to be thrown
```

--- a/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScan.java
+++ b/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScan.java
@@ -80,6 +80,12 @@ public class LanceScan
    */
   private final java.util.Map<String, List<ZoneStats>> zonemapStats;
 
+  /**
+   * Pre-computed surviving fragment IDs from zonemap pruning in LanceScanBuilder. When non-null,
+   * {@link #pruneByZonemapStats} skips re-computing and uses these directly.
+   */
+  private final Set<Integer> cachedSurvivingFragmentIds;
+
   /** Number of partitions after pruning, set during {@link #planInputPartitions()}. */
   private transient int numPartitions = -1;
 
@@ -112,6 +118,7 @@ public class LanceScan
       Filter[] pushedFilters,
       LanceStatistics statistics,
       java.util.Map<String, List<ZoneStats>> zonemapStats,
+      Set<Integer> survivingFragmentIds,
       ZonemapFragmentPruner.PartitionInfo partitionInfo,
       java.util.Map<String, String> initialStorageOptions,
       String namespaceImpl,
@@ -127,6 +134,7 @@ public class LanceScan
         pushedFilters != null ? Arrays.copyOf(pushedFilters, pushedFilters.length) : new Filter[0];
     this.statistics = statistics;
     this.zonemapStats = zonemapStats != null ? zonemapStats : Collections.emptyMap();
+    this.cachedSurvivingFragmentIds = survivingFragmentIds;
     this.partitionInfo = partitionInfo;
     this.initialStorageOptions = initialStorageOptions;
     this.namespaceImpl = namespaceImpl;
@@ -315,17 +323,19 @@ public class LanceScan
    * predicate are skipped entirely, avoiding fragment opens, scan setup, and task scheduling.
    */
   private List<LanceSplit> pruneByZonemapStats(List<LanceSplit> allSplits) {
-    if (zonemapStats.isEmpty()) {
+    // Use cached result from LanceScanBuilder if available, otherwise compute.
+    Set<Integer> allowedIds;
+    if (cachedSurvivingFragmentIds != null) {
+      allowedIds = cachedSurvivingFragmentIds;
+    } else if (!zonemapStats.isEmpty()) {
+      allowedIds = ZonemapFragmentPruner.pruneFragments(pushedFilters, zonemapStats).orElse(null);
+    } else {
       return allSplits;
     }
 
-    java.util.Optional<Set<Integer>> targetFragmentIds =
-        ZonemapFragmentPruner.pruneFragments(pushedFilters, zonemapStats);
-    if (!targetFragmentIds.isPresent()) {
+    if (allowedIds == null) {
       return allSplits;
     }
-
-    Set<Integer> allowedIds = targetFragmentIds.get();
     List<LanceSplit> pruned =
         allSplits.stream()
             .filter(split -> split.getFragments().stream().anyMatch(allowedIds::contains))

--- a/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScan.java
+++ b/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScan.java
@@ -342,7 +342,7 @@ public class LanceScan
             .collect(Collectors.toList());
 
     if (pruned.size() < allSplits.size()) {
-      LOG.info(
+      LOG.debug(
           "Zonemap pruning: {} of {} splits retained," + " allowed fragment IDs: {}",
           pruned.size(),
           allSplits.size(),

--- a/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScanBuilder.java
+++ b/lance-spark-base_2.12/src/main/java/org/lance/spark/read/LanceScanBuilder.java
@@ -164,46 +164,45 @@ public class LanceScanBuilder
             partitionColumn,
             TABLE_OPT_PARTITION_COLUMNS);
       } else {
-        java.util.Optional<Map<Integer, Comparable<?>>> partValues =
-            ZonemapFragmentPruner.computeFragmentPartitionValues(zonemapStats.get(partitionColumn));
-        if (partValues.isPresent()) {
-          partitionInfo =
-              new ZonemapFragmentPruner.PartitionInfo(partitionColumn, partValues.get());
+        Map<Integer, Comparable<?>> partValues =
+            ZonemapFragmentPruner.computeFragmentPartitionValues(zonemapStats.get(partitionColumn))
+                .orElse(null);
+        if (partValues != null) {
+          partitionInfo = new ZonemapFragmentPruner.PartitionInfo(partitionColumn, partValues);
           LOG.info(
               "Detected partition-compatible column '{}' with {} fragments",
               partitionColumn,
-              partValues.get().size());
+              partValues.size());
         }
       }
     }
 
-    // Estimate post-pruning statistics so Spark's JoinSelection can pick BroadcastHashJoin
-    // when the pruned size is below the broadcast threshold. Without this, the full-table size
-    // causes JoinSelection to pick SortMergeJoin, which then gets locked into SPJ — preventing
-    // AQE from switching to broadcast even when one side is tiny after filter pushdown.
-    LanceStatistics statistics;
+    // Pre-compute fragment pruning so we can (a) estimate post-pruning statistics for
+    // JoinSelection (BroadcastHashJoin vs SortMergeJoin) and (b) pass the cached result
+    // to LanceScan to avoid re-computing during planInputPartitions().
+    Set<Integer> survivingFragmentIds = null;
     if (pushedFilters.length > 0 && !zonemapStats.isEmpty()) {
-      java.util.Optional<Set<Integer>> survivingIds =
-          ZonemapFragmentPruner.pruneFragments(pushedFilters, zonemapStats);
-      if (survivingIds.isPresent()) {
-        statistics =
-            LanceStatistics.estimatePostPruning(
-                summary.getTotalRows(),
-                summary.getTotalFilesSize(),
-                summary.getTotalFragments(),
-                survivingIds.get().size());
-        LOG.info(
-            "Estimated post-pruning statistics: {} of {} fragments survive,"
-                + " estimatedSize={}, estimatedRows={} (full: size={}, rows={})",
-            survivingIds.get().size(),
-            summary.getTotalFragments(),
-            statistics.sizeInBytes(),
-            statistics.numRows(),
-            summary.getTotalFilesSize(),
-            summary.getTotalRows());
-      } else {
-        statistics = new LanceStatistics(summary);
-      }
+      survivingFragmentIds =
+          ZonemapFragmentPruner.pruneFragments(pushedFilters, zonemapStats).orElse(null);
+    }
+
+    LanceStatistics statistics;
+    if (survivingFragmentIds != null) {
+      statistics =
+          LanceStatistics.estimatePostPruning(
+              summary.getTotalRows(),
+              summary.getTotalFilesSize(),
+              summary.getTotalFragments(),
+              survivingFragmentIds.size());
+      LOG.debug(
+          "Estimated post-pruning statistics: {} of {} fragments survive,"
+              + " estimatedSize={}, estimatedRows={} (full: size={}, rows={})",
+          survivingFragmentIds.size(),
+          summary.getTotalFragments(),
+          statistics.sizeInBytes(),
+          statistics.numRows(),
+          summary.getTotalFilesSize(),
+          summary.getTotalRows());
     } else {
       statistics = new LanceStatistics(summary);
     }
@@ -223,6 +222,7 @@ public class LanceScanBuilder
         pushedFilters,
         statistics,
         zonemapStats,
+        survivingFragmentIds,
         partitionInfo,
         initialStorageOptions,
         namespaceImpl,
@@ -379,7 +379,7 @@ public class LanceScanBuilder
     }
 
     if (!result.isEmpty()) {
-      LOG.info("Loaded zonemap stats for {} columns: {}", result.size(), result.keySet());
+      LOG.debug("Loaded zonemap stats for {} columns: {}", result.size(), result.keySet());
     }
 
     return result;

--- a/lance-spark-base_2.12/src/test/java/org/lance/spark/read/BaseSpjBroadcastTest.java
+++ b/lance-spark-base_2.12/src/test/java/org/lance/spark/read/BaseSpjBroadcastTest.java
@@ -22,7 +22,6 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
 import java.io.IOException;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.UUID;
 import java.util.stream.Collectors;
@@ -53,9 +52,9 @@ public abstract class BaseSpjBroadcastTest {
   @TempDir Path tempDir;
 
   @BeforeEach
-  public void setup() throws IOException {
+  public void setup() {
     Path rootPath = tempDir.resolve(UUID.randomUUID().toString());
-    Files.createDirectories(rootPath);
+    rootPath.toFile().mkdirs();
     String testRoot = rootPath.toString();
     spark =
         SparkSession.builder()
@@ -76,7 +75,7 @@ public abstract class BaseSpjBroadcastTest {
   }
 
   @AfterEach
-  public void tearDown() {
+  public void tearDown() throws IOException {
     if (spark != null) {
       spark.close();
     }

--- a/lance-spark-base_2.12/src/test/java/org/lance/spark/read/LanceScanTest.java
+++ b/lance-spark-base_2.12/src/test/java/org/lance/spark/read/LanceScanTest.java
@@ -213,6 +213,7 @@ public class LanceScanTest {
             new Filter[0],
             null,
             Collections.emptyMap(),
+            null,
             partInfo,
             Collections.emptyMap(),
             null,

